multiversx/mx-chain-go
# POC: Distribute to waiting from auction based on leaving nodes

## Reasoning behind the pull request
- In case of leaving active nodes in epoch (waiting/eligible), as long as there leaving nodes per shard < number of active nodes && leaving nodes per shard  < number nodes to shuffle per shard, those will not be shuffled out, they can leave. If there are more nodes per shard that are leaving than the number of nodes to shuffle per shard, the remaining will be forced to stay
  
## Proposed changes
- Fix auction list selection algorithm to consider leaving active nodes and forced nodes to stay when computing number of avaiable slots to select qualified nodes.

## Testing procedure
- One easy starting point would be to do something similar to this test: `TestChainSimulator_UnStakeOneActiveNodeAndCheckAPIAuctionList`

Scenario 1: 
- Have every epoch auction list with enough nodes (let's say 8 qualified, 2 unqualified)
- Unstake 1 or more eligible/waiting nodes (no more than all nodes per shard or no more than all eligible) and call auction list api; we should see 8 qualified nodes. Next epoch we need to check that exactly 8 nodes were qualified. Afterwards(some epochs), those leaving nodes will be leaving and replaced by the auction nodes

Scenario 2:
- Have every epoch auction list with MORE nodes (let's say 8 qualified, 20 unqualified)
- Let's say we have number of nodes to shuffle per shard = 2 and 8 eligible nodes + 4 waiting nodes
- Unstake from one or more shards >=2 nodes. We should see in api auction list that we have number of qualified nodes = maxNumNodesFromConfig -  (numRemainingActiveNodes + number of forced nodes to stay); _forced node = every other unstaked node over 2 node unstaked per shard_


**At all times for all these tests, we should always have in system eligible + waiting nodes <= maxNumNodes from config.
Please check this at all times**
